### PR TITLE
Update build files for java8 build

### PR DESCRIPTION
**Pull request by alanpaxton**
For RocksJava 7 we will move from requiring Java 7 to Java 8.

* This simplifies the `Makefile` as we no longer need to deal with Java 7; so we no longer use `javah`.
* Added a java-version target which is invoked by the java target, and which exits if the version of java being used is not 8 or greater.
* Enforces java 8 as a minimum.
* Fixed CMake build.

* Fixed broken java event listener test, as the test was broken and the assertions in the callbacks were not causing assertions in the tests. The callbacks now queue up assertion errors for the main thread of the tests to check.
* Fixed C++ dangling pointers in the test code.

### CODE REVIEW

**Comment by alanpaxton** [comment]
CMake build has passed CI, so I believe this is ready for review.

File: java/CMakeLists.txt [hunk 1]
```diff
@@ -309,13 +309,17 @@ set(JAVA_TESTCLASSPATH ${JAVA_JUNIT_JAR} ${JAVA_HAMCR_JAR} ${JAVA_MOCKITO_JAR} $
 set(JNI_OUTPUT_DIR ${PROJECT_SOURCE_DIR}/java/include)
 file(MAKE_DIRECTORY ${JNI_OUTPUT_DIR})
 
+if(${Java_VERSION_MINOR} VERSION_LESS_EQUAL "7" AND ${Java_VERSION_MAJOR} STREQUAL "1")
```
Comment (adamretter) [inline java/CMakeLists.txt:312]: @alanpaxton Just wondering, as they have changed the version numbering scheme after Java 1.8, does this still work on say Java 11? How about Java 17 ;-)

**Review by adamretter — COMMENTED** [review]
LGTM - just one question about version checks when Java 17 is used (as it is semantically similar to Java 1.7)

File: java/CMakeLists.txt (hunk 1 above)
Comment (alanpaxton) [reply java/CMakeLists.txt:312]: The values come from CMake's FindJava plugin. I just went and played with it. Java 11, 13, 15, 17 all have correct ${Java_VERSION_MAJOR} of 11,. 13, 15, 17 respectively. Java 1.6, 1.7 and 1.8 have ${Java_VERSION_MAJOR} STREQUAL "1" so I claim this test works as intended.

**Review by adamretter — APPROVED** [review]
It's ready to go :-)

**Pull request closed without merging**